libs/api/database/src/lib/content/stores/tags.store.ts

```diff
@@ -0,0 +1,167 @@
+import { BadRequestException, Injectable, Logger, NotFoundException } from '@nestjs/common';
+import { InjectModel } from '@nestjs/mongoose';
+import { Model } from 'mongoose';
+import { TagsDocument } from '../schemas';
+import { TagKind, TagsForm } from '@dragonfish/shared/models/content';
+import * as sanitize from 'sanitize-html';
+import { isNullOrUndefined } from '@dragonfish/shared/functions';
+import { TagsTree } from '@dragonfish/shared/models/content/tags.model';
+
+@Injectable()
+export class TagsStore {    
+    private readonly logger: Logger = new Logger(TagsStore.name);
+
+    constructor(
+        @InjectModel('Tags') private readonly tags: Model<TagsDocument>,
+    ) {}
+
+    /**
+     * Returns all tags of a specified kind, sorted alphabetically.
+     * @param kind
+     */
+    async fetchTags(kind: TagKind): Promise<TagsDocument[]> {
+        return this.tags.find({ kind: kind }).sort({ name: 1 });
+    }
+
+    /**
+     * Returns a single tag with the given ID, or null if not found.
+     * @param id The ID of the tag to find.
+     * @returns The tag with the ID.
+     */
+    async findOne(id: string): Promise<TagsDocument> {
+        return this.tags.findById(id);
+    }
+
+    /**
+     * Returns the given tag and all its descendants.
+     * @param king
+     * @param tagId
+     */
+    async fetchDescendants(tagId: string): Promise<TagsTree> {
+        const parentTag = await this.tags.findById(tagId);
+        if (!parentTag) {
+            throw new NotFoundException("No tag with that ID found.");
+        }
+
+        const root: TagsTree = {
+            ...parentTag.toObject(),
+            children: []
+        };
+        const tree = await this.buildDescendantTree(root);
+
+        return tree;
+    }
+
+    private async buildDescendantTree(tagNode: TagsTree): Promise<TagsTree> {
```

Comment:
aaaaa recursion, aaaa

Reply:
For the case of someone getting access to a mod account and using it to just add a bunch of child tags, or some other scenario I'm not thinking of, we should perhaps limit it to just three levels. I don't think we have plans to go further than that.

Reply:
This code checks out. At first I thought there was an issue, but I initially missed that the child.children was being set in the loop, and not the child itself. My notes posted below for others to look at if they don't understand.

a
b c
d e | f g

1
Input: a
currentRoot.children[b].children = bDT(b).children
currentRoot.children[c].children = bDT(c).children
return currentRoot

2
Input: b
currentRoot.children[d].children = bDT(d).children
currentRoot.children[e].children = bDT(e).children
return currentRoot

3
Input: d
d has null children
return d

2
Input: b
currentRoot.children[d].children = d.children = null
currentRoot.children[e].children = bDT(e).children
return currentRoot

4
Input: e
e has null children
return e

2
Input: b
currentRoot.children[d].children = d.children = null
currentRoot.children[e].children = e.children = null
return currentRoot // b - d, e

1
Input: a
currentRoot.children[b].children = (b - d, e).children = [d, e]
currentRoot.children[c].children = bDT(c).children
return currentRoot

Reply:
lgtm